lib/rust/flexer/src/group.rs

```diff
@@ -10,6 +10,150 @@ pub mod rule;
+    ///
+    /// It returns the identifier of the newly-created group.
+    pub fn define_group(&mut self, name:String, parent_index:Option<usize>) -> Identifier {
+        let id = self.next_id().into();
+        let group = Group::new(id,name,parent_index);
+        self.groups.push(group);
+        id
+    }
+
+    /// Adds an existing `group` to the registry, updating and returning its identifier.
+    pub fn add_group(&mut self, mut group:Group) -> Identifier {
+        let new_id = self.next_id().into();
+        group.id = new_id;
+        self.groups.push(group);
+        new_id
+    }
+
+    /// Creates a rule that matches `pattern` for the group identified by `group_id`.
+    ///
+    /// Panics if `group_id` refers to a nonexistent group.
+    pub fn create_rule(&mut self, group:Identifier, pattern:&Pattern, callback:&str) {
+        let err = format!("The provided group_id {:?} is invalid.",group);
+        let group = self.group_from_id_mut(group).expect(&err);
+        group.create_rule(pattern,callback);
+    }
+
+    /// Associates the provided `rule` with the group identified by `group_id`.
+    ///
+    /// Panics if `group_id` refers to a nonexistent group.
+    pub fn add_rule(&mut self, group:Identifier, rule:Rule) {
+        let err = format!("The provided group_id {:?} is invalid.",group);
+        let group = self.group_from_id_mut(group).expect(&err);
+        group.add_rule(rule);
+    }
+
+    /// Collates the entire set of rules that are matchable when the lexer has the group identified
+    /// by `group_id` as active.
+    ///
+    /// This set of rules includes the rules inherited from any parent groups.
+    pub fn rules_for(&self, group:Identifier) -> Option<Vec<&Rule>> {
+        self.group_from_id(group).map(|group| {
+            let mut parent = group.parent_index.and_then(|ix|self.group_from_id(ix.into()));
+            let mut rules = (&group.rules).iter().collect_vec();
+            while let Some(parent_group) = parent {
+                if parent_group.id == group.id {
+                    panic!("There should not be cycles in parent links for lexer groups.")
+                }
+                rules.extend((&parent_group.rules).iter());
+                parent = parent_group.parent_index.and_then(|ix|self.group_from_id(ix.into()));
+            }
+
+            rules
+        })
+    }
+
+    /// Obtains a reference to the group for the given `group_id`.
+    pub fn group_from_id(&self, group:Identifier) -> Option<&Group> {
+        self.groups.get(group.val)
+    }
+
+    /// Obtains a mutable reference to the group for the given `group_id`.
+    pub fn group_from_id_mut(&mut self, group:Identifier) -> Option<&mut Group> {
+        self.groups.get_mut(group.val)
+    }
+
+    /// Converts the group identified by `group_id` into an NFA.
+    ///
+    /// Returns `None` if the group does not exist, or if the conversion fails.
+    pub fn to_nfa_from(&self, group:Identifier) -> Option<NFA> {
```

Comment:
I owuld name it `to_nfa`. Right now "to nfa from <group>" is for me like converting this group to nfa only, while in fact we are using all connected groups to convert to nfa.

Reply:
No we're not. We're converting the single group only.